harpoon/internal/testing/panic.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+package testing
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+)
+
+const godogFailMessage = "FailNow or SkipNow called"
+
+type stringer interface {
+	String() string
+}
+
+func WrapWithPanicHandler[U any](needsLog bool, exitBehavior ExitBehavior, fn U) U {
+	fnValue := reflect.ValueOf(fn)
+	return reflect.MakeFunc(fnValue.Type(), func(args []reflect.Value) (results []reflect.Value) {
+		runTestingFunc := func(fn func(t *TestingT)) {
+			if len(args) > 0 {
+				if ctx, ok := args[0].Interface().(context.Context); ok {
+					t, ok := ctx.Value(testingContextKey).(*TestingT)
+					if ok && t != nil {
+						fn(t)
+					}
+				}
+			}
+		}
+
+		defer func() {
+			if err := recover(); err != nil {
+				// swallow any panics from godog
+				message, ok := isGodogLogPanic(err)
+				if !ok {
+					// otherwise re-panic
+					panic(err)
+				}
+				runTestingFunc(func(t *TestingT) {
+					if priorError := t.PriorError(); priorError != "" {
+						message = priorError
+					}
+				})
+				switch exitBehavior {
+				case ExitBehaviorNone:
+				case ExitBehaviorLog:
+					fmt.Print(message)
+				case ExitBehaviorTerminateProgram:
+					fmt.Print(message)
+					os.Exit(1)
+				case ExitBehaviorTestFail:
+					select {
+					case TerminationChan <- TerminationError{
+						Message:  message,
+						NeedsLog: needsLog,
+					}:
+					default:
```

Comment:
Why a non-blocking send here? Could `TerminationChan` be a buffered channel with size one instead?

Reply:
the non-blocking send is because of the fact that we only read once in the top-level channel reader, the

```go
			select {
			case <-ctx.Done():
				return
			case errMessage = <-internaltesting.TerminationChan:
			}
```

code.

Since all of that only fails the test after the suite is fully run, rescuing from a `FailNow`-based panic here could mean that multiple function calls in cleanup hooks (where termination conditions actually matter since everything else is handled inline by godog checks on the tainted internal error tracking) might also call `FailNow` and any subsequent panic would cause this write to block. If we used a buffered channel we'd still want the non-blocking send here just because we have no idea how many cleanup functions might potentially `FailNow`.